tests/qnodes/test_qnode_passthru.py

```diff
@@ -0,0 +1,186 @@
+    DTYPE = tf.float64 if tf else None
+
+    def test_arraylike_args(self, tensornet_tf_device, tol):
+        """Tests that PassthruQNode can use array-like TF objects as positional arguments."""
+
+        def circuit(x):
+            qml.RX(x[0], wires=[0])
+            qml.RX(2*x[1], wires=[1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+        x = tf.Variable([1.1, 1.4], dtype=self.DTYPE)
+        res = node(x)
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype == self.DTYPE
+
+    def test_arraylike_keywordargs(self, tensornet_tf_device, tol):
+        """Tests that qnodes use array-like TF objects as keyword-only arguments."""
+
+        def circuit(*, x=None):
+            qml.RX(x[0], wires=[0])
+            qml.RX(2*x[1], wires=[1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+        x = tf.Variable([1.1, 1.4], dtype=self.DTYPE)
+        res = node(x=x)
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype == self.DTYPE
+
+    def test_tensor_operations(self, tensornet_tf_device, tol):
+        """Tests the evaluation of a PassthruQNode involving algebraic operations between tensor parameters,
+        and TF functions acting on them."""
+
+        def circuit(phi, theta):
+            x = phi * theta
+            qml.RX(x[0], wires=0)
+            qml.RY(tf.cos(x[1]), wires=1)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.Hadamard(1))
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+
+        phi = tf.Variable(np.array([0.7, -1.2]), dtype=self.DTYPE)
+        theta = tf.Variable(1.7, dtype=self.DTYPE)
+        res = node(phi, theta)
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype == self.DTYPE
+
+    def test_evaluate(self, tensornet_tf_device, tol):
+        """Tests the evaluation of a PassthruQNode."""
+
+        def circuit(x):
+            qml.RX(x[0], wires=0)
+            qml.RY(x[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.RX(x[2], wires=1)
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        x = np.array([-1.4, 0.2, 2.1])
+        x_tf = tf.Variable(x, dtype=self.DTYPE)
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+        res = node(x_tf)
+
+        # compare to a reference node
+        ref_device = qml.device('default.qubit', wires=2)
+        ref_node = BaseQNode(circuit, ref_device)
+        ref_res = ref_node(x)
+
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype == self.DTYPE
+        assert res.numpy() == pytest.approx(ref_res, abs=tol)
```

Comment:
First time I've seen `pytest.approx()` used. What makes this more appropriate from the established pattern with `np.allclose()`?

Reply:
`pytest.approx` seems to be roughly as frequent in the tests as `np.allclose`. It should be the same as `np.allclose(..., atol=tol, rtol=0)`. I suppose it's a bit shorter and a part of the pytest framework which we use anyway.

https://docs.pytest.org/en/latest/reference.html#pytest-approx